File: Core/src/main/java/com/inverce/mod/processing/Job.java
package com.inverce.mod.processing;

import com.inverce.mod.core.IM;
import com.inverce.mod.core.functional.IConsumer;

class Job<ITEM, RESULT> implements IConsumer<ProcessingQueue> {
    ITEM item;
    Processor<ITEM, RESULT> processor;

    public Job(ITEM item, Processor<ITEM, RESULT> processor) {
        this.item = item;
        this.processor = processor;
    }

    @Override
    public void consume(ProcessingQueue queue) {
        try {
            RESULT result = processor.processJob(item);
            IM.onBg().execute(() -> queue.finishJob(new JobResult<>(this, result)));
        } catch (Exception ex) {
            IM.onBg().execute(() -> queue.finishJob(new JobResult<>(this, ex)));
        }
    }
}

File: Processing/src/main/java/com/inverce/mod/v2/processing/Job.kt
package com.inverce.mod.v2.processing

class Job<T, R>(
        val processor: JobProcessor<T, R>,
        val steps: List<T>
) {
    val results: List<R> = emptyList()


}


class Processors {
    fun <T, R> create(process: (item: T, context: JobContext<T, R>) -> R) {


    }


}

typealias JobProcessor<T, R> = (item: T, context: JobContext<T, R>) -> R


val pr: JobProcessor<Int, Double> = { i, c -> i.toLong() }
                .then { i, c -> i.toDouble()

                }


                .then(object : JobProcessor<Long, Double> {
                    override fun process(context: JobContext<Long, Double>, item: Long): Double {
                        return item.toDouble()
                    }
                })

fun <T, B, R> JobProcessor<T, B>.then(processor: JobProcessor<B, R>): JobProcessor<T, R> {

}

interface JobContext<T, R> {
    fun reportProgress(progress: Float)
    fun reportResult(result: R)
    fun reportResult(result: Exception)
}

class JobResult<ITEM, RESULT>(
        val job: Job<ITEM, RESULT>,
        val result: RESULT?,
        val exception: Exception?)
